### PR TITLE
Select and deselect all functionality added

### PR DESCRIPTION
**Pull request by sd1998**
Fix to issue #1727.

### CODE REVIEW

**Comment by sd1998** [comment]
@davigonz @jesmrec @theScrabi I have implemented the feature.

**Comment by theScrabi** [comment]
@sd1998 @davigonz Do we need the deselect option? It basically does the same as the back arrow button.

out

**Comment by davigonz** [comment]
> Do we need the deselect option? It basically does the same as the back arrow button.

I see ... yes, is basically doing the same. What do you think @theScrabi and @sd1998 about including a "Select inverse" option instead of "Deselect all" as you can see in the next gif from other app?

jun-18-2018 13-56-52

This is a nice option in my opinion, pretty useful when you have a scenario like the one above, in which the user wants to select all the files except a couple of them.

**Comment by sd1998** [comment]
@davigonz I agree with @theScrabi regarding the fact that the deselect option in not required because of the presence of the back button and the Idea that you have suggested regarding select inverse is very useful. Should I implement that instead?

**Comment by jesmrec** [comment]
- Deselecting is removing all selected items in multiselection mode, but maybe i want to select other ones. 
- Back arrow is cancelling multiselection mode.

For me, these are very different things.

**Comment by davigonz** [comment]
> Deselecting is removing all selected items in multiselection mode, but maybe i want to select other ones.
If the user wants to select more along with the ones already selected, he won't press the "Deselecting all"  button.

> Back arrow is cancelling multiselection mode.
As we can see in the gif attached in https://github.com/owncloud/android/pull/2242#issuecomment-397971204, back arrow is deselecting all the files previously selected so is the same behaviour as when you press the "Deselect all" button.

**Comment by jesmrec** [comment]
> As we can see in the gif attached in #2242 (comment), back arrow is deselecting all the files previously selected so is the same behaviour as when you press the "Deselect all" button.

this is the point. The back arrow is used to cancel the multiselection mode. Deselecting means (imho), uncheck all checkboxes that are checked, not cancelling multiselection mode. If i mistake in the files i want to select to be removed (or any other action), and i want to clean my selection, i should use the "Deselect" option. If i do not want to select multiple files, i tap on the back arrow.

**Comment by davigonz** [comment]
> I agree with @theScrabi regarding the fact that the deselect option in not required because of the presence of the back button and the Idea that you have suggested regarding select inverse is very useful. Should I implement that instead?

Think so, go for it 😉

**Comment by sd1998** [comment]
@davigonz I have changes deselect all to select inverse.

File: src/com/owncloud/android/files/FileMenuFilter.java [hunk 1]
```diff
@@ -90,15 +90,15 @@ public FileMenuFilter(OCFile targetFile, Account account, ComponentsGetter cg,
      *
      * @param menu              Options or context menu to filter.
      */
-    public void filter(Menu menu) {
+    public void filter(Menu menu,boolean displaySelectAll,boolean displayDeselectAll) {
```
Comment (dutta14) [inline src/com/owncloud/android/files/FileMenuFilter.java]: Minor: checkstyle - spaces between parameters

File: src/com/owncloud/android/files/FileMenuFilter.java [hunk 2]
```diff
@@ -90,15 +90,15 @@ public FileMenuFilter(OCFile targetFile, Account account, ComponentsGetter cg,
...
         if (mFiles == null || mFiles.size() <= 0) {
             hideAll(menu);
 
         } else {
             List<Integer> toShow = new ArrayList<>();
             List<Integer> toHide = new ArrayList<>();
 
-            filter(toShow, toHide);
+            filter(toShow, toHide,displaySelectAll,displayDeselectAll);
```
Comment (dutta14) [inline src/com/owncloud/android/files/FileMenuFilter.java]: Minor: checkstyle - spaces between parameters

File: src/com/owncloud/android/files/FileMenuFilter.java [hunk 3]
```diff
@@ -137,7 +137,7 @@ private void hideAll(Menu menu) {
      * @param toShow            List to save the options that must be shown in the menu.
      * @param toHide            List to save the options that must be shown in the menu.
      */
-    private void filter(List<Integer> toShow, List <Integer> toHide) {
+    private void filter(List<Integer> toShow, List <Integer> toHide,boolean displaySelectAll,boolean displayDeselectAll) {
```
Comment (dutta14) [inline src/com/owncloud/android/files/FileMenuFilter.java]: Minor: checkstyle - spaces between parameters

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 4]
```diff
@@ -542,6 +547,9 @@ public boolean onOptionsItemSelected(MenuItem item) {
 
 
         switch (item.getItemId()) {
+            case R.id.action_select_all: {
```
Comment (dutta14) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: `case` should not have braces.

File: src/com/owncloud/android/ui/fragment/FileDetailFragment.java [hunk 5]
```diff
@@ -225,7 +225,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                 mContainerActivity,
                 getActivity()
             );
-            mf.filter(menu);
+            mf.filter(menu,false,false);
         }
```
Comment (dutta14) [inline src/com/owncloud/android/ui/fragment/FileDetailFragment.java]: Minor: checkstyle - space between parameters

File: src/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 6]
```diff
@@ -527,8 +535,7 @@ public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
                 mContainerActivity,
                 getActivity()
             );
-            mf.filter(menu);
-
+            mf.filter(menu,mEnableSelectAll,true);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/fragment/OCFileListFragment.java]: Minor: checkstyle - space between parameters

File: src/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 7]
```diff
@@ -778,6 +785,12 @@ private boolean fileIsDownloading(OCFile file) {
                 ((FileActivity) mContainerActivity).getAccount(), file);
     }
 
+    public void selectAll(){
+        for(int i = 0;i < mAdapter.getCount();i++){
+            getListView().setItemChecked(i,true);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/fragment/OCFileListFragment.java]: Minor: spaces between parameters and loop conditions, and opening braces.

File: src/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 8]
```diff
@@ -828,6 +842,17 @@ public boolean onFileActionChosen(int menuId) {
 
         /// actions possible on a batch of files
         switch (menuId) {
+            case R.id.file_action_select_all: {
+                selectAll();
```
Comment (dutta14) [inline src/com/owncloud/android/ui/fragment/OCFileListFragment.java]: Minor: `case` should not have braces.

File: src/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 9]
```diff
@@ -828,6 +842,17 @@ public boolean onFileActionChosen(int menuId) {
...
+
+                return true;
+            }
+            case R.id.action_deselect_all: {
+                for(int i = 0;i < mAdapter.getCount();i++){
```
Comment (dutta14) [inline src/com/owncloud/android/ui/fragment/OCFileListFragment.java]: `case` should not have braces.  Loop checkstyle: put spaces at proper places.

File: src/com/owncloud/android/ui/preview/PreviewAudioFragment.java [hunk 10]
```diff
@@ -325,7 +325,7 @@ public void onPrepareOptionsMenu(Menu menu) {
             mContainerActivity,
             getActivity()
         );
-        mf.filter(menu);
+        mf.filter(menu,false,false);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/preview/PreviewAudioFragment.java]: Checkstyle: put spaces between parameters

File: src/com/owncloud/android/ui/preview/PreviewImageFragment.java [hunk 11]
```diff
@@ -261,7 +261,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                 mContainerActivity,
                 getActivity()
             );
-            mf.filter(menu);
+            mf.filter(menu,false,false);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/preview/PreviewImageFragment.java]: Checkstyle: put spaces between parameters

File: src/com/owncloud/android/ui/preview/PreviewTextFragment.java [hunk 12]
```diff
@@ -298,7 +298,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                     mContainerActivity,
                     getActivity()
             );
-            mf.filter(menu);
+            mf.filter(menu,false,false);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/preview/PreviewTextFragment.java]: Checkstyle: put spaces between parameters

File: src/com/owncloud/android/ui/preview/PreviewVideoFragment.java [hunk 13]
```diff
@@ -353,7 +353,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                 mContainerActivity,
                 getActivity()
         );
-        mf.filter(menu);
+        mf.filter(menu,false,false);
```
Comment (dutta14) [inline src/com/owncloud/android/ui/preview/PreviewVideoFragment.java]: Checkstyle: put spaces between parameters

**Comment by davigonz** [comment]
Thanks a lot @dutta14 for reviewing this PR 😊

**Comment by sd1998** [comment]
@davigonz I have made the changes.

File: res/values/strings.xml [hunk 14]
```diff
@@ -408,8 +408,11 @@
     <string name="log_send_mail_subject">%1$s Android app logs</string>
     <string name="log_progress_dialog_text">Loading data &#8230;</string>
 
+    <string name="actionbar_select_all">Select all</string>
+    <string name="actionbar_select_inverse">Select inverse</string>
+
     <string name="saml_authentication_required_text">Authentication required</string>
-    <string name="saml_authentication_wrong_pass">Wrong password</string>
+    <string name="saml_authentication_wrong_pass">W ong password</string>
```
Comment (davigonz) [inline res/values/strings.xml]: There's a typo in here **W ong password**

File: src/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 15]
```diff
@@ -48,6 +48,7 @@
 import android.support.v4.app.FragmentTransaction;
 import android.support.v4.content.ContextCompat;
 import android.support.v4.content.LocalBroadcastManager;
+import android.util.Log;
```
Comment (davigonz) [inline src/com/owncloud/android/ui/activity/FileDisplayActivity.java]: This is not used

**Comment by davigonz** [comment]
@sd1998 just a couple of things more before starting with QA

**Comment by sd1998** [comment]
@davigonz I have made the changes.

**Comment by davigonz** [comment]
Code approved, ready to pass QA

**Comment by jesmrec** [comment]
Let's QA this

**Comment by davigonz** [comment]
> What do you think about removing the options if the folder is empty?

yes, it makes sense

**Comment by jesmrec** [comment]
@sd1998 go ahead!

**Comment by jesmrec** [comment]
@sd1998 how is going on? if you need some help with this (or any other) contribution, do not hesitate to ping us.

**Comment by pixolex** [comment]
And select interval?

**Comment by sd1998** [comment]
@jesmrec Sorry for the delay I will start working on this now.

**Comment by pixolex** [comment]
Select the first file, select the last file and after that have a
button/option to select all the files between this two.

A segunda, 13/08/2018, 07:21, Jesús Recio Rincón <notifications@github.com>
escreveu:

**Comment by jesmrec** [comment]
I understand your point, but i am not sure about the UX. You can upload pics here by dragging and dropping.

@sd1998 no worries, take your time :)

**Comment by pixolex** [comment]
screenshot_20180813-081549__01

**Comment by davigonz** [comment]
> i get your idea, but i think it should be addressed in other issue as an improvement to the current one, so we can keep the incremental development. What do you think @sd1998?

+1, I would address this in a separate issue

**Comment by sd1998** [comment]
@davigonz @jesmrec I have made the changes.

**Comment by davigonz** [comment]
Merged, thanks a lot for this contribution @sd1998 , it will be included in 2.9 version

**Pull request merged**